/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
import pandas as pd

df = pd.read_excel('Online retail.xlsx')
display(df.head())

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                             burgers,meatballs,eggs                                                                                                                                                                             
1                                            chutney                                                                                                                                                                             
2                                     turkey,avocado                                                                                                                                                                             
3  mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                     low fat yogurt

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
print(df.isnull().sum())

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil    0
dtype: int64


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
print("Number of duplicate rows:", df.duplicated().sum())
df.drop_duplicates(inplace=True)
print("Number of duplicate rows after removal:", df.duplicated().sum())

Number of duplicate rows: 2325
Number of duplicate rows after removal: 0


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
transactions = df.values.tolist()
transaction_list = [transaction[0].split(',') for transaction in transactions if transaction[0] is not None]
all_items = sorted(list(set([item for sublist in transaction_list for item in sublist])))

encoded_transactions = []
for transaction in transaction_list:
    encoded_row = [0] * len(all_items)
    for item in transaction:
        try:
            item_index = all_items.index(item)
            encoded_row[item_index] = 1
        except ValueError:
            # Handle items not in all_items (shouldn't happen if all_items is comprehensive)
            pass
    encoded_transactions.append(encoded_row)

df_encoded = pd.DataFrame(encoded_transactions, columns=all_items)
display(df_encoded.head())

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0           0        0                  0          0        0            0   
1           0        0                  0          0        0            0   
2           0        0                  0          0        1            0   
3           0        0                  0          0        0            0   
4           0        0                  0          0        0            0   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0      0               0          0            0  ...       0               0   
1      0               0          0            0  ...       0               0   
2      0               0          0            0  ...       1               0   
3      0               0          0            0  ...       0               0   
4      0               0          0            0  ...       0               0   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0            0           0                 0                  0   
1            0           0                 0                  0   
2            0           0                 0                  0   
3            0           0                 0                  0   
4            0           0                 0                  0   

   whole wheat rice  yams  yogurt cake  zucchini  
0                 0     0            0         0  
1                 0     0            0         0  
2                 0     0            0         0  
3                 1     0            0         0  
4                 0     0            0         0  

[5 rows x 120 columns]

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
from mlxtend.frequent_patterns import apriori

frequent_itemsets = apriori(df_encoded, min_support=0.01, use_colnames=True)
display(frequent_itemsets.head())

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

support             itemsets
0  0.029179            (almonds)
1  0.011014  (antioxydant juice)
2  0.045797            (avocado)
3  0.012560              (bacon)
4  0.015459     (barbecue sauce)

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
from mlxtend.frequent_patterns import association_rules

rules = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.1)
display(rules.head())

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

antecedents      consequents  antecedent support  consequent support  \
0   (almonds)  (mineral water)            0.029179            0.299710   
1   (avocado)      (chocolate)            0.045797            0.205217   
2   (avocado)   (french fries)            0.045797            0.192657   
3   (avocado)           (milk)            0.045797            0.170048   
4   (avocado)  (mineral water)            0.045797            0.299710   

    support  confidence      lift  representativity  leverage  conviction  \
0  0.010821    0.370861  1.237399               1.0  0.002076    1.113092   
1  0.010242    0.223629  1.089716               1.0  0.000843    1.023715   
2  0.011594    0.253165  1.314069               1.0  0.002771    1.081019   
3  0.010821    0.236287  1.389528               1.0  0.003034    1.086732   
4  0.015845    0.345992  1.154421               1.0  0.002120    1.070766   

   zhangs_metric   jaccard  certainty  kulczynski  
0       0.197619  0.034022   0.101602    0.203483  
1       0.086281  0.042536   0.023165    0.136767  
2       0.250476  0.051107   0.074947    0.156673  
3       0.293786  0.052780   0.079810    0.149962  
4       0.140185  0.048066   0.066089    0.199430

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
display(rules.head(10))

print("\nInterpretation of a few rules:")
print("Rule 1: {} -> {}".format(list(rules.iloc[0]['antecedents']), list(rules.iloc[0]['consequents'])))
print("Support: {:.4f}, Confidence: {:.4f}, Lift: {:.4f}".format(rules.iloc[0]['support'], rules.iloc[0]['confidence'], rules.iloc[0]['lift']))
print("Interpretation: This rule indicates that customers who bought {} are likely to also buy {}. The confidence of {:.4f} means that {:.2f}% of transactions containing {} also contain {}. A lift of {:.4f} (greater than 1) suggests that the purchase of {} and {} together is more likely than the individual probabilities of purchasing them, indicating a positive association.".format(list(rules.iloc[0]['antecedents']), list(rules.iloc[0]['consequents']), rules.iloc[0]['confidence'], rules.iloc[0]['confidence']*100, list(rules.iloc[0]['antecedents']), list(rules.iloc[0]['consequents']), rules.iloc[0]['lift'], list(rules.iloc[0]['antecedents']), list(rules.iloc[0]['consequents'])))

print("\nRule 2: {} -> {}".format(list(rules.iloc[1]['antecedents']), list(rules.iloc[1]['consequents'])))
print("Support: {:.4f}, Confidence: {:.4f}, Lift: {:.4f}".format(rules.iloc[1]['support'], rules.iloc[1]['confidence'], rules.iloc[1]['lift']))
print("Interpretation: This rule suggests that customers who bought {} are likely to also buy {}. The confidence of {:.4f} means that {:.2f}% of transactions containing {} also contain {}. A lift of {:.4f} (greater than 1) indicates a positive association between the purchase of {} and {}.".format(list(rules.iloc[1]['antecedents']), list(rules.iloc[1]['consequents']), rules.iloc[1]['confidence'], rules.iloc[1]['confidence']*100, list(rules.iloc[1]['antecedents']), list(rules.iloc[1]['consequents']), rules.iloc[1]['lift'], list(rules.iloc[1]['antecedents']), list(rules.iloc[1]['consequents'])))

print("\nRule 3: {} -> {}".format(list(rules.iloc[2]['antecedents']), list(rules.iloc[2]['consequents'])))
print("Support: {:.4f}, Confidence: {:.4f}, Lift: {:.4f}".format(rules.iloc[2]['support'], rules.iloc[2]['confidence'], rules.iloc[2]['lift']))
print("Interpretation: This rule shows that customers who bought {} tend to also buy {}. The confidence of {:.4f} indicates that {:.2f}% of transactions with {} also include {}. A lift of {:.4f} (greater than 1) implies that the co-occurrence of {} and {} is more frequent than expected by random chance.".format(list(rules.iloc[2]['antecedents']), list(rules.iloc[2]['consequents']), rules.iloc[2]['confidence'], rules.iloc[2]['confidence']*100, list(rules.iloc[2]['antecedents']), list(rules.iloc[2]['consequents']), rules.iloc[2]['lift'], list(rules.iloc[2]['antecedents']), list(rules.iloc[2]['consequents'])))

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

antecedents      consequents  antecedent support  consequent support  \
0   (almonds)  (mineral water)            0.029179            0.299710   
1   (avocado)      (chocolate)            0.045797            0.205217   
2   (avocado)   (french fries)            0.045797            0.192657   
3   (avocado)           (milk)            0.045797            0.170048   
4   (avocado)  (mineral water)            0.045797            0.299710   
5   (avocado)      (spaghetti)            0.045797            0.229565   
6  (brownies)           (eggs)            0.045024            0.208116   
7  (brownies)   (french fries)            0.045024            0.192657   
8  (brownies)  (mineral water)            0.045024            0.299710   
9  (brownies)      (spaghetti)            0.045024            0.229565   

    support  confidence      lift  representativity  leverage  conviction  \
0  0.010821    0.370861  1.237399               1.0  0.002076    1.113092   
1  0.010242    0.223629  1.089716               1.0  0.000843    1.023715   
2  0.011594    0.253165  1.314069               1.0  0.002771    1.081019   
3  0.010821    0.236287  1.389528               1.0  0.003034    1.086732   
4  0.015845    0.345992  1.154421               1.0  0.002120    1.070766   
5  0.011594    0.253165  1.102800               1.0  0.001081    1.031599   
6  0.010628    0.236052  1.134231               1.0  0.001258    1.036567   
7  0.011208    0.248927  1.292074               1.0  0.002534    1.074920   
8  0.013913    0.309013  1.031039               1.0  0.000419    1.013463   
9  0.010628    0.236052  1.028255               1.0  0.000292    1.008490   

   zhangs_metric   jaccard  certainty  kulczynski  
0       0.197619  0.034022   0.101602    0.203483  
1       0.086281  0.042536   0.023165    0.136767  
2       0.250476  0.051107   0.074947    0.156673  
3       0.293786  0.052780   0.079810    0.149962  
4       0.140185  0.048066   0.066089    0.199430  
5       0.097691  0.043956   0.030631    0.151835  
6       0.123925  0.043825   0.035277    0.143560  
7       0.236708  0.049488   0.069698    0.153551  
8       0.031524  0.042056   0.013284    0.177717  
9       0.028774  0.040264   0.008419    0.141174


Interpretation of a few rules:
Rule 1: ['almonds'] -> ['mineral water']
Support: 0.0108, Confidence: 0.3709, Lift: 1.2374
Interpretation: This rule indicates that customers who bought ['almonds'] are likely to also buy ['mineral water']. The confidence of 0.3709 means that 37.09% of transactions containing ['almonds'] also contain ['mineral water']. A lift of 1.2374 (greater than 1) suggests that the purchase of ['almonds'] and ['mineral water'] together is more likely than the individual probabilities of purchasing them, indicating a positive association.

Rule 2: ['avocado'] -> ['chocolate']
Support: 0.0102, Confidence: 0.2236, Lift: 1.0897
Interpretation: This rule suggests that customers who bought ['avocado'] are likely to also buy ['chocolate']. The confidence of 0.2236 means that 22.36% of transactions containing ['avocado'] also contain ['chocolate']. A lift of 1.0897 (greater than 1) indicates a positive association between the purchase of ['avocado'] and ['chocolate'].

Rule

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

## Summary:

### Data Analysis Key Findings

*   The dataset initially contained 2325 duplicate rows, which were successfully removed.
*   No missing values were found in the dataset.
*   The data was successfully transformed into a one-hot encoded transactional format suitable for the Apriori algorithm.
*   The Apriori algorithm identified frequent itemsets with a minimum support of 0.01.
*   Association rules were generated from the frequent itemsets with a minimum confidence of 0.1.
*   Analysis of the top rules revealed positive associations between certain products, such as 'almonds' and 'mineral water' (confidence: 37.09%, lift: 1.24), 'avocado' and 'chocolate' (confidence: 22.36%, lift: 1.09), and 'avocado' and 'french fries' (confidence: 25.32%, lift: 1.31).

### Insights or Next Steps

*   The identified association rules can be used to inform product placement strategies in a physical store or recommendations in an online store to encourage cross-selling.
*   Further analysis could involve exploring rules with higher confidence or lift values, or segmenting customers to find associations specific to different groups.


In [ ]:
Interview Questions.
1.What is lift and why is it important in Association rules?
Ans:Lift is a measure used in Association rule mining to determine the strength of an association rule between two itemssets.
Lift is important to make it much more reliable indicator of a true association than other metrics like confidence and support .


2.What is support and confidence. How do u calculate them?
Ans:Support and Confidence are two fundamental metrics used to evaluate the strength and usefulness of a discovered rule.
In association rule mining, support and confidence are two key mertics use dto evaluate the strength of an association rule.


3.What are some limitations or challenges of Association rules mining?
Ans:Association rule mining, despite its utility in finding interesting patterns in large datasets faces several significant limitations and challenges.
>The "Too Many Rules " Problem.
>The "No Casuality" Issue.
>Sensitivity to Thresholds.
>Computational Expense.